src/spearman.jl

```diff
@@ -0,0 +1,169 @@
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+export CorrelationTest, SpearmanCorrelationTest
+
+abstract CorrelationTest <: HypothesisTest
+
+"Sum squared difference of ranks (midranks for ties)"
+spearman_S(xrank, yrank) = sumabs2(xrank .- yrank)
+
+immutable SpearmanCorrelationTest <: CorrelationTest
+    # Tied ranking for x and y
+    xrank::Vector{Float64}
+    yrank::Vector{Float64}
+    # Adjustment for ties
+    xtiesadj::Float64
+    ytiesadj::Float64
+    # Sum squared difference of ranks
+    S::Float64
+    # Number of points
+    n::Int
+    # Spearman's ρ
+    ρ::Float64
+
+    function SpearmanCorrelationTest(x, y)
+
+        n = length(x)
+        (n != length(y)) && throw(ErrorException("x and y must have the same length"))
+
+        xrank, xtiesadj = HypothesisTests.tiedrank_adj(x)
+        yrank, ytiesadj = HypothesisTests.tiedrank_adj(y)
+
+        S = spearman_S(xrank, yrank)
+
+        ρ = corspearman(x, y)
+
+        new(xrank, yrank, xtiesadj, ytiesadj, S, n, ρ)
+    end
+
+end
+
+testname(::SpearmanCorrelationTest) = "Spearman's rank correlation test"
+
+# parameter of interest: name, value under h0, point estimate
+population_param_of_interest(x::SpearmanCorrelationTest) = ("Spearman's ρ", 0.0, x.ρ)
+
+function show_params(io::IO, x::SpearmanCorrelationTest, ident)
+    println(io, ident, "Number of points:                    ", x.n)
+    println(io, ident, "Spearman's ρ:                        ", x.ρ)
+    println(io, ident, "S (Sum squared difference of ranks): ", x.S)
+    println(io, ident, "adjustment for ties in x:            ", x.xtiesadj)
+    println(io, ident, "adjustment for ties in y:            ", x.ytiesadj)
+end
+
+function P_from_null_S_values(S_null, x::SpearmanCorrelationTest, tail)
+    S_null_mean = mean(S_null)
+    # S is approximately normally distributed
+    # S and ρ are inversely proportional
+    S_null[:] = S_null .- S_null_mean # center
+    S_centered = x.S - S_null_mean    # center
+    if tail == :both
+        modS = abs(S_centered)
+        mean(S_null .<= -modS) + mean(S_null .>= modS)
+    elseif tail == :right
+        mean(S_null .<= S_centered)
+    elseif tail == :left
+        mean(S_null .>= S_centered)
+    else
+        throw(ArgumentError("tail=$(tail) is invalid"))
+    end
+end
+
+function spearman_P_exact(x::SpearmanCorrelationTest, tail)
+    S_null = Float64[ spearman_S(perm, x.yrank) for perm in permutations(x.xrank) ]
+    P_from_null_S_values(S_null, x, tail)
+end
+
+function spearman_P_sampling(x::SpearmanCorrelationTest, tail)
+    # 360000 samples gives an se(P) < 0.0005 for P < 0.1
+    X = copy(x.xrank)
+    S_null = Float64[ spearman_S(shuffle!(X), x.yrank) for sample in 1:360000 ]
```

Comment:
I don't think it's a problem to set the seed for a whole test session. Other uses of `rand` should do the same anyway to prevent random failures, so they shouldn't be affected.